src/net_processing.cpp

```diff
@@ -4719,6 +4719,16 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         LogPrint(BCLog::NET, "received block %s peer=%d\n", pblock->GetHash().ToString(), pfrom.GetId());
 
+        const CBlockIndex* prev_block{WITH_LOCK(m_chainman.GetMutex(), return m_chainman.m_blockman.LookupBlockIndex(pblock->hashPrevBlock))};
+
+        if (IsBlockMutated(/*block=*/*pblock,
```

Comment:
should we be checking PoW first before doing potentially a lot of hashing?

Reply:
Ok this doesn't seem to matter: we do this hashing even for already-known blocks we get sent unilaterally, and `MAX_PROTOCOL_MESSAGE_LENGTH` bounds the size of hashing to something reasonable.

Reply:
49257c0304828a185c273fcb99742c54bbef0c8e I had the same question, specifically why we're not doing this after the `CalculateHeadersWork()` check below, and right before `ProcessBlock`.

A code comment would be useful here.